docs/design2.md

```diff
@@ -0,0 +1,109 @@
+# Key Transparency 2.0
+
+## Objective
+
+Key Transparency currently suffers from a design-space trade-off in two dimensions between
+low latency key updates and verification bandwidth; each snapshot of the
+key-value Merkle tree creates an additional point in time that clients must
+audit when inspecting their account history.
+Using Key Transparency v0.4.0 and below requires the presence of powerful
+auditors who can verify that no information has been lost between all snapshots.
+
+| Update Latency | History Audit    | Key Lookup |
+|----------------|------------------|------------|
+| hr             | 45Kb/day         | ~2Kb       |
+| min            | 2Mb/day          | ~2Kb       |
+| s              | 150Mb/day        | ~2Kb       |
+
+
+Key Transparency 2.0 is a new set of algorithms that removes this fundamental trade-off by sharing
+the task of auditing between pairs of clients. As a result, all configurations have efficiently
+auditable data structures that no not require the presence of 3rd party auditors.
+
+We expect that many applications will want to prioritize lower latency and
+reliability by using a *trust, then verify* mode whereby clients receive key
+updates that they can later verify as being part of the global, consistent
+state.  However, this does not preclude individual clients that have a lower
+tolerance for risk from waiting for full, globally consistent verification
+before using fresh public keys from their peers.
+
+| Update Latency | History Audit    | Key Lookup |
+|----------------|------------------|------------|
+| hr             | 18Kb O(1)        | ~2Kb       |
+| min            | 30Kb O(1)        | ~2Kb       |
+| s              | 40Kb O(1)        | ~2Kb       |
+
+
+## Data Structures
+
+### Gossip Network
+The job of the gossip network is to ensure that there is a single, globally consistent, lineage of log roots.
+To keep ![n^2](https://render.githubusercontent.com/render/math?math=n%5E2) communication costs low, and to prevent sybil attacks, clients use a small ~20 set of gossip nodes.
+
+Each gossip node fetches the latest signed log root (SLR) and verifies
+consistency with all previously seen roots with a single consistency proof.
+After verifying, the gossip node signs the log root.
+
+Gossip nodes offer the following APIs:
+
+1. Get Signed Root
+
+### Log of Map Roots
+
+The log server collects signatures from the gossip nodes to produce a Signed Log Root that has a quorum of gossip signatures.
+
+The log server contains a list of map roots representing sequential snapshots of a key-value directory.
+
+Log servers offer the following APIs:
+
+1. Latest Signed Log Root
+1. Get consistency proof between log size a and b.
+1. Get log item i with inclusion proof to log size a.
+
+### Map Root Snapshots
+Each Map Root represents a snapshot of a key-value dictionary.
+The map is implemented as a sparse Merkle tree of fixed depth.
```

Comment:
Is the "fixed depth" required here? This is how it was implemented in Trillian Map, but a Patricia trie, for instance, doesn't have a fixed depth and provides similar properties.

It feels like the depth fixedness is implied from VRF (if it outputs fixed length blobs), but not required.

Reply:
Fixed depth is certainly not required. I actually would prefer a Patricia Tri. A previous version of the document called out migrating to a Patrica Tri as a TODO, but I thought it would be clearer to not include TODOs in this design doc.

Reply:
Sure, I'm saying that this doc shouldn't require a fixed depth either. I suggest dropping " of fixed depth" from the sentence.